src/state/safety.rs

```diff
@@ -0,0 +1,315 @@
+//! [BlockTree]. This generally requires that the block or nudge is well-formed, and that the quroum
+//! certificate that it references is cryptographically correct and safe. However, since nudges
+//! implement a version of the non-pipelined HotStuff consensus for validator-set-updating blocks, a
+//! non-decide-phase nudge is only safe if it is broadcasted in the next view relative to the view in
+//! which its justify is collected. 
+//! 
+//! While [safe_block], [safe_nudge], and [safe_qc] define what it means for a block, nudge, or QC,
+//! to be safe, and hence be allowed to trigger state updates, [qc_to_lock] and [block_to_commit]
+//! can be used to determine which updates can be made on seeing a safe nudge or a safe block.
+//! This depends on the commit and locking rules.
+//! 
+//! ## Safe QC
+//! 
+//! A safe QC must match the chain id of the replica (to avoid having state updates triggered by a
+//! QC from another blockchain). It must also have an appropriate phase depending on whether the
+//! block it justifies has associated validator set updates or not, and it must be for a known
+//! block, i.e., a block that is already stored in the BlockTree.
+//! 
+//! Most importantly, it must either:
+//! 1. Reference a block that extends the branch of the block referenced by the locked QC, or
+//! 2. Have a higher view than the locked QC.
+//! 
+//! The former condition guarantees safety in case a quorum of replicas is locked on a given block,
+//! while the latter ensures liveness in case the quorum has moved to a conflicting branch and
+//! there exists evidence for it. The evidence would be a newely received QC with a view higher than the
+//! view of the replica's locked QC.
+//! 
+//! ## Commit rules
+//! 
+//! To avoid inconsistent state due to Byzantine leaders causing safety-threatening branch switches, the
+//! pipelined HotStuff protocol requires that in order to be committed a block must be followed by a
+//! 3-chain. Concretely, the three QCs referenced by its child, grandchild, and great-grandchild, 
+//! serving as its prepareQC, precommitQC, and commitQC respectively, must have consecutive views.
+//! [block_to_commit] enforces this rule for the pipelined protocol for non-validator-set-updating
+//! blocks. It is not required in [safe_block] since blocks can be safely added even if their 
+//! ancestors don't have consecutive views - but commit of the branch will only be triggered once
+//! there is a 3-chain.
+//! 
+//! To obtain an equivalent guarantee for the non-pipelined protocol for validator-set-updating blocks,
+//! we require that the prepareQC, precommitQC, and commitQC for a validator-set-updating block received
+//! via nudges have consecutive views. If this flow is interrupted by a temporary loss of synchrony or
+//! a Byzantine leader, the validator-set-updating block has to be re-proposed, as specified in
+//! [repropose_block]. This rule is enforced through [safe_nudge]. Unlike block insertions, nudges
```

Comment:
Technically, we do not **have** to repropose the original validator set updating block. We can instead create a new block that is a "sibling" to the original validator set updating block and propose that instead, correct?

Reply:
Not sure if I understand, but what would be the advantage of that? With the current implementation if a vs-updating block B has already been inserted by the leader and other replicas then (1) the leader doesn't have to call `produce_block` to re-propose it, and (2) other replicas can vote for the re-proposal without having to insert it again (recall that now they can vote for a block that's already in their block tree, they just don't insert it again). 

And why produce a sibling - if the block is the same, then the hash would be the same, and everyone already has a block with this hash in their block tree, if not then they insert.

Reply:
I just wanted to point this out. The current behavior (re-proposing an existing block) is fine, though.